File: packages/pigeon/platform_tests/flutter_null_safe_unit_tests/test/all_datatypes_test.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:typed_data';

import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_unit_tests/all_datatypes.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';
import 'all_datatypes_test.mocks.dart';

@GenerateMocks(<Type>[BinaryMessenger])
void main() {
  test('with null values', () async {
    final Everything everything = Everything();
    final BinaryMessenger mockMessenger = MockBinaryMessenger();
    when(mockMessenger.send('dev.flutter.pigeon.HostEverything.echo', any))
        .thenAnswer((Invocation realInvocation) async {
      const StandardMessageCodec codec = StandardMessageCodec();
      final Object input =
          codec.decodeMessage(realInvocation.positionalArguments[1]);
      return codec.encodeMessage(<String, Object>{'result': input});
    });
    final HostEverything api = HostEverything(binaryMessenger: mockMessenger);
    final Everything result = await api.echo(everything);
    expect(result.aBool, isNull);
    expect(result.anInt, isNull);
    expect(result.aDouble, isNull);
    expect(result.aString, isNull);
    expect(result.aByteArray, isNull);
    expect(result.a4ByteArray, isNull);
    expect(result.a8ByteArray, isNull);
    expect(result.aFloatArray, isNull);
    expect(result.aList, isNull);
    expect(result.aMap, isNull);
  });

  test('with values', () async {
    final Everything everything = Everything();
    everything.aBool = false;
    everything.anInt = 123;
    everything.aDouble = 2.0;
    everything.aString = 'hello';
    everything.aByteArray = Uint8List.fromList(<int>[1, 2, 3, 4]);
    everything.a4ByteArray = Int32List.fromList(<int>[1, 2, 3, 4]);
    everything.a8ByteArray = Int64List.fromList(<int>[1, 2, 3, 4]);
    everything.aFloatArray =
        Float64List.fromList(<double>[1.0, 2.5, 3.0, 4.25]);
    everything.aList = <int>[1, 2, 3, 4];
    everything.aMap = <String, int>{'hello': 1234};
    final BinaryMessenger mockMessenger = MockBinaryMessenger();
    when(mockMessenger.send('dev.flutter.pigeon.HostEverything.echo', any))
        .thenAnswer((Invocation realInvocation) async {
      const StandardMessageCodec codec = StandardMessageCodec();
      final Object input =
          codec.decodeMessage(realInvocation.positionalArguments[1]);
      return codec.encodeMessage(<String, Object>{'result': input});
    });
    final HostEverything api = HostEverything(binaryMessenger: mockMessenger);
    final Everything result = await api.echo(everything);
    expect(result.aBool, everything.aBool);
    expect(result.anInt, everything.anInt);
    expect(result.aDouble, everything.aDouble);
    expect(result.aString, everything.aString);
    expect(result.aByteArray, everything.aByteArray);
    expect(result.a4ByteArray, everything.a4ByteArray);
    expect(result.a8ByteArray, everything.a8ByteArray);
    expect(result.aFloatArray, everything.aFloatArray);
    expect(result.aList, everything.aList);
    expect(result.aMap, everything.aMap);
  });
}

File: packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/AllDatatypesTest.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import <Flutter/Flutter.h>
#import <XCTest/XCTest.h>
#import "EchoMessenger.h"
#import "all_datatypes.gen.h"

///////////////////////////////////////////////////////////////////////////////////////////
@interface AllDatatypesTest : XCTestCase
@end

///////////////////////////////////////////////////////////////////////////////////////////
@implementation AllDatatypesTest

- (void)testAllNull {
  Everything *everything = [[Everything alloc] init];
  EchoBinaryMessenger *binaryMessenger =
      [[EchoBinaryMessenger alloc] initWithCodec:FlutterEverythingGetCodec()];
  FlutterEverything *api = [[FlutterEverything alloc] initWithBinaryMessenger:binaryMessenger];
  XCTestExpectation *expectation = [self expectationWithDescription:@"callback"];
  [api echoEverything:everything
           completion:^(Everything *_Nonnull result, NSError *_Nullable error) {
             XCTAssertNil(error);
             XCTAssertNotNil(result);
             XCTAssert([result.aBool isEqual:[NSNull null]]);
             XCTAssert([result.anInt isEqual:[NSNull null]]);
             XCTAssert([result.aDouble isEqual:[NSNull null]]);
             XCTAssert([result.aString isEqual:[NSNull null]]);
             XCTAssert([result.aByteArray isEqual:[NSNull null]]);
             XCTAssert([result.a4ByteArray isEqual:[NSNull null]]);
             XCTAssert([result.a8ByteArray isEqual:[NSNull null]]);
             XCTAssert([result.aFloatArray isEqual:[NSNull null]]);
             XCTAssert([result.aList isEqual:[NSNull null]]);
             XCTAssert([result.aMap isEqual:[NSNull null]]);
             [expectation fulfill];
           }];
  [self waitForExpectations:@[ expectation ] timeout:1.0];
}

- (void)testAllEquals {
  Everything *everything = [[Everything alloc] init];
  everything.aBool = @NO;
  everything.anInt = @(1);
  everything.aDouble = @(2.0);
  everything.aString = @"123";
  everything.aByteArray = [FlutterStandardTypedData
      typedDataWithBytes:[@"1234" dataUsingEncoding:NSUTF8StringEncoding]];
  everything.a4ByteArray = [FlutterStandardTypedData
      typedDataWithInt32:[@"1234" dataUsingEncoding:NSUTF8StringEncoding]];
  everything.a8ByteArray = [FlutterStandardTypedData
      typedDataWithInt64:[@"12345678" dataUsingEncoding:NSUTF8StringEncoding]];
  everything.aFloatArray = [FlutterStandardTypedData
      typedDataWithFloat64:[@"12345678" dataUsingEncoding:NSUTF8StringEncoding]];
  everything.aList = @[ @(1), @(2) ];
  everything.aMap = @{@"hello" : @(1234)};
  everything.mapWithObject = @{@"hello" : @(1234), @"goodbye" : @"world"};
  EchoBinaryMessenger *binaryMessenger =
      [[EchoBinaryMessenger alloc] initWithCodec:FlutterEverythingGetCodec()];
  FlutterEverything *api = [[FlutterEverything alloc] initWithBinaryMessenger:binaryMessenger];
  XCTestExpectation *expectation = [self expectationWithDescription:@"callback"];
  [api echoEverything:everything
           completion:^(Everything *_Nonnull result, NSError *_Nullable error) {
             XCTAssertEqual(result.aBool, everything.aBool);
             XCTAssertEqual(result.anInt, everything.anInt);
             XCTAssertEqual(result.aDouble, everything.aDouble);
             XCTAssertEqualObjects(result.aString, everything.aString);
             XCTAssertEqualObjects(result.aByteArray.data, everything.aByteArray.data);
             XCTAssertEqualObjects(result.a4ByteArray.data, everything.a4ByteArray.data);
             XCTAssertEqualObjects(result.a8ByteArray.data, everything.a8ByteArray.data);
             XCTAssertEqualObjects(result.aFloatArray.data, everything.aFloatArray.data);
             XCTAssertEqualObjects(result.aList, everything.aList);
             XCTAssertEqualObjects(result.aMap, everything.aMap);
             XCTAssertEqualObjects(result.mapWithObject, everything.mapWithObject);
             [expectation fulfill];
           }];
  [self waitForExpectations:@[ expectation ] timeout:1.0];
}

@end
